pkg/networkserver/downlink.go

```diff
@@ -573,17 +571,49 @@ type scheduledDownlink struct {
 
 type downlinkSchedulingError []error
 
-func (errs downlinkSchedulingError) IsRetryable() bool {
+func (errs downlinkSchedulingError) Error() string {
+	return errSchedule.Error()
+}
+
+// gatewayErrors returns gateway errors represented by errs and boolean
+// indicating whether all errors in errs represent gateway errors.
+func (errs downlinkSchedulingError) gatewayErrors() ([]error, bool) {
+	gtwErrs := make([]error, 0, len(errs))
+	allGatewayErrors := true
+	for _, gsErr := range errs {
+		ttnGSErr, ok := errors.From(gsErr)
+		if !ok {
+			allGatewayErrors = false
+			continue
+		}
+		for _, d := range ttnGSErr.Details() {
+			gtwErr, ok := d.(error)
```

Comment:
In which case does this cast fail?

Reply:
No clue, details are not type checked - they're just `interface{}` (https://github.com/TheThingsNetwork/lorawan-stack/blob/8102bd25586153ce41207714555402277ee84fbd/pkg/errors/details.go#L33-L56), so there are no guarantees on the type. Better safe than sorry.
@htdvisser any comment on how is this supposed to be used?

Reply:
But isn't the details than a `ttnpb` type? Or (also) an `error`?

Reply:
According to https://github.com/TheThingsNetwork/lorawan-stack/blob/8102bd25586153ce41207714555402277ee84fbd/pkg/errors/details.go#L47
it's
> Usually structs from ttnpb or google.golang.org/genproto/googleapis/rpc/errdetails.

so there's no guarantee

@htdvisser can we refactor these to use `error` instead of `interface{}`?

Reply:
Error details wraps existing gRPC API. There is absolutely no guarantee that any of these actually implement the `error` interface. The details is just a list of proto messages. When we send errors over gRPC they contain a single `ttnpb.ErrorDetails` that carries ttn-lw-stack specific details of the error like the error namespace and name. This could be extended with gRPC (googleapis) error details such as field violation details, debug info, localized messages, retry info, etc. but I think we'll use our own protos instead.

Considering that `WithDetails` is currently only used in two places in master (and incorrectly by the looks of it) I think it may be better to completely remove the `Details` funcs from the `errors` package until we can actually use it correctly.

Reply:
How would that work? These can come from any valid API implementation, right? I think we need proper type assertion.

My comment is mostly about directly type casting to `error`, which I don't think is going to work, but not sure.